AppTaskQueue/appscale/taskqueue/queue.py

```diff
@@ -235,7 +235,538 @@ def __repr__(self):
+    """
+    super(PostgresPullQueue, self).__init__(queue_info, app)
+    self.pg_connection = pg_connection
+    self.pg_cursor = pg_connection.cursor()
+    self.ensure_tables_created()
+
+  @property
+  def tasks_table_name(self):
+    return 'pullqueue-{}'.format(self.name)
+
+  def ensure_tables_created(self):
+    try:
+      self.pg_cursor.execute(
+        'CREATE TABLE IF NOT EXISTS "{table_name}" ('
+        '  task_name varchar(500) NOT NULL,'
+        '  time_enqueued timestamp NOT NULL,'
+        '  retry_count integer NOT NULL,'
+        '  lease_expires timestamp NOT NULL,'
+        '  payload text,'
+        '  tag varchar(500),'
+        '  PRIMARY KEY (task_name)'
+        ');'
+        'CREATE INDEX "{table_name}-eta-retry-tag-index" '
+        '  ON "{table_name}" USING BTREE (lease_expires, retry_count, tag);'
+        'CREATE INDEX "{table_name}-retry-eta-tag-index" '
+        '  ON "{table_name}" (retry_count, lease_expires, tag);'
+        .format(table_name=self.tasks_table_name)
+      )
+      self.pg_connection.commit()
+    except Exception as err:
+      logger.error('Rolling back transaction ({err})'.format(err=err))
+      self.pg_connection.rollback()
+      raise
+
+  def add_task(self, task):
+    """ Adds a task to the queue.
+
+    Args:
+      task: A Task object.
+    Raises:
+      InvalidTaskInfo if the task ID already exists in the queue
+        or it doesn't have payloadBase64 attribute.
+    """
+    import psycopg2
+    if not hasattr(task, 'payloadBase64'):
+      raise InvalidTaskInfo('{} is missing a payload.'.format(task))
+
+    enqueue_time = datetime.datetime.utcnow()
+    try:
+      lease_expires = task.leaseTimestamp
+    except AttributeError:
+      lease_expires = datetime.datetime.utcfromtimestamp(0)
+
+    try:
+      task.queueName = self.name
+      task.enqueueTimestamp = enqueue_time
+      task.leaseTimestamp = lease_expires
+
+      self.pg_cursor.execute(
+        'INSERT INTO "{table}" ( '
+        '  task_name, payload, time_enqueued, '
+        '  lease_expires, retry_count, tag '
+        ')'
+        'VALUES ( '
+        '  %(task_name)s, %(payload)s, %(time_enqueued)s, '
+        '  %(lease_expires)s, %(retry_count)s, %(tag)s '
+        ')'.format(table=self.tasks_table_name),
+        vars={
+          'task_name': task.id,
+          'payload': task.payloadBase64,
+          'time_enqueued': enqueue_time,
+          'lease_expires': lease_expires,
+          'retry_count': 0,
+          'tag': getattr(task, 'tag', None)
+        }
+      )
+
+      self.pg_connection.commit()
+      logger.debug('Added task: {}'.format(task))
+
+    except psycopg2.IntegrityError as err:
+      name_taken_msg = 'Task name already taken: {}'.format(task.id)
+      logger.warning('{msg}. Rolling back transaction({err})'
+                     .format(msg=name_taken_msg, err=err))
+      self.pg_connection.rollback()
+      raise InvalidTaskInfo(name_taken_msg)
```

Comment:
To match GAE behavior (which we don't do in the Cassandra implementation), we would need to differentiate between an existing task and a tombstoned task name. I'm fine with leaving that detail for later, though.

Reply:
Thanks. As differentiating errors would require changes in http handlers, I would implement it as a separate PR.